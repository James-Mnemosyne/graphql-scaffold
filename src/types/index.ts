export enum ResolverType {
  Query = 'Query',
  Mutation = 'Mutation',
  Field = 'Field',
  Unknown = 'Unknown',
}

export interface Config {
  baseFilePath: string;
  resolverFilePath: string;
  resolverTestFilePath: string;
  authorizerFilePath: string;
  resolverType: string;
  resolverName: string;
  schemaIndexFilePath: string;
  schemaFilePath: string;
}

export enum FixtureNodeVariation {
  BaseFixtureNode = 'BaseFixtureNode',
  EnumFixtureNode = 'EnumFixtureNode',
  ListFixtureNode = 'ListFixtureNode',
  FunctionFixtureNode = 'FunctionFixtureNode',
  NonNullNode = 'NonNullNode',
  ObjectFixtureNode = 'ObjectFixtureNode',
  RootFixtureNode = 'RootFixtureNode',
  ScalarFixtureNode = 'ScalarFixtureNode',
  UnionFixtureNode = 'UnionFixtureNode',
  UnknownFixtureNode = 'UnknownFixtureNode',
}

type FixtureAttributes = {
  name: string;
  defined?: boolean;
};

export type BaseFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.BaseFixtureNode;
  fieldVariations: any;
  resolved: boolean;
};

export type ScalarFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.ScalarFixtureNode;
  scalarName: string;
};

export type ListFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.ListFixtureNode;
  child: FixtureNode;
};

export type ObjectFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.ObjectFixtureNode;
  fields: FixtureNode[];
};

export type EnumFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.EnumFixtureNode;
  keys: string[];
};

export type UnionFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.UnionFixtureNode;
  elements: FixtureNode[];
};

export type FunctionFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.FunctionFixtureNode;
  inputs: FixtureNode[];
  response: FixtureNode;
};

export type RootFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.RootFixtureNode;
  self: FixtureNode;
};

export type UnknownFixtureNode = FixtureAttributes & {
  variation: FixtureNodeVariation.UnknownFixtureNode;
  location: string;
};

export type FixtureNode =
  | BaseFixtureNode
  | EnumFixtureNode
  | ListFixtureNode
  | ObjectFixtureNode
  | ScalarFixtureNode
  | UnionFixtureNode
  | FunctionFixtureNode
  | UnknownFixtureNode;
