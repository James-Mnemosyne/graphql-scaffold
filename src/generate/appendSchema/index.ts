import * as fs from 'fs';
import { appendImports } from '../../utils/appendImports';
import { Config } from '../../types';

function getSchemaIndexFlieBody(config) {
  const resolverName = config.resolverName;

  return `import ${resolverName} from './schema.graphql';

export { ${resolverName} };
`;
}

function addSchemaIndexFile(config: Config) {
  const fileBody = getSchemaIndexFlieBody(config);
  fs.writeFileSync(config.schemaIndexFilePath, fileBody);
}

export function appendSchema(config: Config): void {
  appendImports(config.schemaFilePath, config);
  addSchemaIndexFile(config);
}
