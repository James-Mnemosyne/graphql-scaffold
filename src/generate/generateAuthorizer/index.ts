import * as fs from 'fs';
import { appendImports } from '../../utils/appendImports';
import { Config } from '../../types';

function getAuthorizerBody(config) {
  const resolverName = config.resolverName;
  const lowercaseResolverType = config.resolverType.toLowerCase();

  return `import { AuthenticationContextType } from 'src/graphql-retail/middleware/authenticate/types';

/**
 * Authorizer for ${resolverName} ${lowercaseResolverType}
 */
export async function ${resolverName}(context: AuthenticationContextType) {
  // DEAR CODE REVIEWER,
  // IF YOU ARE SEEING THIS, YOU SHOULD NOT ALLOW THIS THROUGH CODE REVIEW.
  // THIS IS AN AUTO-GENERATED FUNCTION
  // AND SHOULD BE REPLACED WITH VALID AUTHENTICATION LOGIC.
};
`;
}

export function generateAuthorizer(config: Config): void {
  appendImports(config.authorizerFilePath, config);

  const resolverBody = getAuthorizerBody(config);
  const fileName = config.authorizerFilePath;
  fs.writeFileSync(fileName, resolverBody);
}
