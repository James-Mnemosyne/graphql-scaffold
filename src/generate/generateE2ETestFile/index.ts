import { Config } from '../../types';
import { writeFileWithRecursiveDirCreation } from '../../utils/writeFileWithRecursiveDirCreation';

function getE2ETestBody(config: Config, fixture: Object) {
  const resolverName = config.resolverName;

  return `import { getTestUserToken } from 'src/graphql-retail/testing/utils/getTestUserToken';
  import { makeGraphQLRequest } from 'src/graphql-retail/testing/utils/makeGraphQLRequest';
  import gql from 'graphql-tag';
  
  const query = gql\`
    query ${resolverName} {
      ${resolverName} {
        __typename
      }
    }
  \`;
  
  // This is an e2e test for ${resolverName}, that verifies that it progressese through the entire server, with no issues
  describe('${resolverName}', () => {
    it('fetches correctly', async () => {
      const userToken = (await getTestUserToken({ email: 'usere2etestemail+titan123@gmail.com' })).token;
      const result = await makeGraphQLRequest({ query, variables: ${JSON.stringify(
        fixture
      )}, headers: { authorization: \`Bearer \${userToken}\` } });
  
      expect(result.data).toMatchSnapshot();
      expect(result.errors).toEqual(undefined);
    }, 25000);
  });
  
`;
}

// TODO (orange): re-enable ability to parse schema for better handling.
export function generateE2ETestFile(config: Config, fixture: Object): void {
  const resolverTestBody = getE2ETestBody(config, fixture);
  const fileName = config.resolverE2ETestFilePath;
  writeFileWithRecursiveDirCreation(fileName, resolverTestBody);
}
