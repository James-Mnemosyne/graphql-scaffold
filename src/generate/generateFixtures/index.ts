import { FixtureNode, FixtureNodeVariation } from './types';
import { generateFixturesForEnumFixtureNode } from './generateFixturesForEnumFixtureNode';
import { generateFixturesForFunctionFixtureNode } from './generateFixturesForFunctionFixtureNode';
import { generateFixturesForListFixtureNode } from './generateFixturesForListFixtureNode';
import { generateFixturesForObjectFixtureNode } from './generateFixturesForObjectFixtureNode';
import { generateFixturesForScalarNode } from './generateFixturesForScalarNode';
import { generateFixturesForUnionNode } from './generateFixturesForUnionNode';
import { compress } from './utils';

export function generateFixtures(
  node: FixtureNode,
  fixtureNodes: Record<string, FixtureNode>,
  maxFixtures: number
): any[] {
  switch (node.variation) {
    case FixtureNodeVariation.ScalarFixtureNode:
      const scalarResults = generateFixturesForScalarNode(node, fixtureNodes, maxFixtures);
      return compress(scalarResults, maxFixtures);
    case FixtureNodeVariation.UnionFixtureNode:
      const unionResults = generateFixturesForUnionNode(node, fixtureNodes, maxFixtures);
      return compress(unionResults, maxFixtures);
    case FixtureNodeVariation.FunctionFixtureNode:
      const functionResults = generateFixturesForFunctionFixtureNode(node, fixtureNodes, maxFixtures);
      return compress(functionResults, maxFixtures);
    case FixtureNodeVariation.ObjectFixtureNode:
      const objectResults = generateFixturesForObjectFixtureNode(node, fixtureNodes, maxFixtures);
      return compress(objectResults, maxFixtures);
    case FixtureNodeVariation.EnumFixtureNode:
      const enumResults = generateFixturesForEnumFixtureNode(node);
      return compress(enumResults, maxFixtures);
    case FixtureNodeVariation.ListFixtureNode:
      const listResults = generateFixturesForListFixtureNode(node, fixtureNodes, maxFixtures);
      return compress(listResults, maxFixtures);
    default:
      throw new Error(`Unsupported generic fixture type generation for ${node.variation}.`);
  }
}
