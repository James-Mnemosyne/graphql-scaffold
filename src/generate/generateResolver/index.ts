import * as fs from 'fs';
import { appendImports } from '../../utils/appendImports';
import { Config, ResolverType } from '../../types';

function getResolverTypeName(resolverType: ResolverType) {
  return resolverType === ResolverType.Mutation ? 'Mutations' : resolverType === ResolverType.Query ? 'Queries' : '';
}

function getResolverBody(config) {
  const resolverName = config.resolverName;
  const lowercaseResolverType = config.resolverType.toLowerCase();
  const resolverTypeName = getResolverTypeName(config.resolverType);

  return `import { ${resolverTypeName} } from 'graphql-retail/types';

/**
 * ${resolverName} ${lowercaseResolverType}
 */
export const ${resolverName}: ${resolverTypeName}['${resolverName}'] = async function (
  _parent,
  input,
  _context
) {
  // TODO implement
  console.log('input', input);
  throw new Error('Unsupported ${lowercaseResolverType} ${resolverName}.')
};
`;
}

export function generateResolver(config: Config): void {
  appendImports(config.resolverFilePath, config);

  const resolverBody = getResolverBody(config);
  const fileName = config.resolverFilePath;
  fs.writeFileSync(fileName, resolverBody);
}
