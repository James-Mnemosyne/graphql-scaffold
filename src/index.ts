#!/usr/bin/env node

import { generate } from './generate';
import { getConfig } from './getConfig';

async function run() {
  const config = await getConfig();
  await generate(config);
}

run()
  .then(() => {
    process.exit(0);
  })
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
