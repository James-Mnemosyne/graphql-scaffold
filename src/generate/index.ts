import { runBeforeCommands } from '../commands/runBeforeCommands';
import { runAfterCommands } from '../commands/runAfterCommands';
import { Config } from '../types';
import { generateAuthorizer } from './generateAuthorizer';
import { generateResolver } from './generateResolver';
import { generateResolverTestFile } from './generateResolverTestFile';
import { generateE2ETestFile } from './generateE2ETestFile';
import { generateFixture } from './generateTypes';

export async function generate(config: Config) {
  console.log('Starting.');
  console.log('Running before commands.');
  await runBeforeCommands(config);
  console.log('Before commands complete.');
  generateAuthorizer(config);
  console.log(`Generated authorizer file in ${config.authorizerFilePath}.`);
  generateResolver(config);
  console.log(`Generated resolver file in ${config.resolverFilePath}.`);
  const fixture = await generateFixture(config);
  console.log(`Generated fixture for tests.`);
  generateResolverTestFile(config, fixture);
  console.log(`Generated resolver test file in ${config.resolverTestFilePath}.`);
  generateE2ETestFile(config, fixture);
  console.log(`Generated e2e test file in ${config.resolverE2ETestFilePath}.`);
  console.log('Running after commands.');
  await runAfterCommands(config);
  console.log('After commands complete.');
  console.log('Finished.');
}
