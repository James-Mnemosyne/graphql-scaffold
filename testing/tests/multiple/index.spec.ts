import { generate } from '../../../src/generate';
import * as fs from 'fs';
import { Config, ResolverType } from '../../../src/types';

const resolverName0 = 'thing0';
const resolverName1 = 'thing1';
const extensionPath = 'queries/multiple';
const baseFilePath = `testing/results/multiple/graphql`;

const config0: Config = {
  beforeCommands: [],
  afterCommands: [],
  baseFilePath: `${baseFilePath}`,
  resolverFilePath: `${baseFilePath}/resolvers/${extensionPath}/${resolverName0}/index.ts`,
  resolverTestFilePath: `${baseFilePath}/resolvers/${extensionPath}/${resolverName0}/index.spec.ts`,
  resolverE2ETestFilePath: `${baseFilePath}/e2eTests/${extensionPath}/${resolverName0}/index.spec.ts`,
  authorizerFilePath: `${baseFilePath}/authorizers/${extensionPath}/${resolverName0}/index.ts`,
  resolverType: ResolverType.Query,
  resolverName: resolverName0,
  schemaFilePath: `${baseFilePath}/schemas/${extensionPath}/${resolverName0}/schema.graphql`,
  testType: 'spec',
};

const config1: Config = {
  beforeCommands: [],
  afterCommands: [],
  baseFilePath: `${baseFilePath}`,
  resolverFilePath: `${baseFilePath}/resolvers/${extensionPath}/${resolverName1}/index.ts`,
  resolverTestFilePath: `${baseFilePath}/resolvers/${extensionPath}/${resolverName1}/index.spec.ts`,
  resolverE2ETestFilePath: `${baseFilePath}/e2eTests/${extensionPath}/${resolverName1}/index.spec.ts`,
  authorizerFilePath: `${baseFilePath}/authorizers/${extensionPath}/${resolverName1}/index.ts`,
  resolverType: ResolverType.Query,
  resolverName: resolverName1,
  schemaFilePath: `${baseFilePath}/schemas/${extensionPath}/${resolverName1}/schema.graphql`,
  testType: 'spec',
};

const schemaFileContents0 = `type Query {
  thing0: Int!
}
`;
const schemaFileContents1 = `type Query {
  thing1: Int!
}
`;

describe('Query', () => {
  it('with only basic object', async () => {
    if (fs.existsSync('testing/results/multiple')) {
      fs.rmdirSync('testing/results/multiple', { recursive: true });
    }
    fs.mkdirSync(`testing/results/multiple/graphql/schemas/queries/multiple/${resolverName0}`, { recursive: true });
    fs.writeFileSync(
      `testing/results/multiple/graphql/schemas/queries/multiple/${resolverName0}/schema.graphql`,
      schemaFileContents0
    );
    fs.mkdirSync(`testing/results/multiple/graphql/schemas/queries/multiple/${resolverName1}`, { recursive: true });
    fs.writeFileSync(
      `testing/results/multiple/graphql/schemas/queries/multiple/${resolverName1}/schema.graphql`,
      schemaFileContents1
    );

    await generate(config0);
    await generate(config1);

    expect(fs.existsSync(config0.authorizerFilePath)).toBeTruthy();
    expect(fs.existsSync(config1.authorizerFilePath)).toBeTruthy();
    expect(fs.existsSync('testing/results/multiple/graphql/authorizers/queries/multiple/index.ts')).toBeTruthy();
    expect(fs.existsSync('testing/results/multiple/graphql/authorizers/queries/index.ts')).toBeTruthy();
    expect(fs.existsSync('testing/results/multiple/graphql/authorizers/index.ts')).toBeTruthy();
    expect(fs.existsSync(config0.resolverFilePath)).toBeTruthy();
    expect(fs.existsSync(config1.resolverFilePath)).toBeTruthy();
    expect(fs.existsSync(config0.resolverTestFilePath)).toBeTruthy();
    expect(fs.existsSync(config1.resolverTestFilePath)).toBeTruthy();
    expect(fs.existsSync('testing/results/multiple/graphql/resolvers/queries/multiple/index.ts')).toBeTruthy();
    expect(fs.existsSync('testing/results/multiple/graphql/resolvers/queries/index.ts')).toBeTruthy();

    expect(fs.readFileSync(config0.authorizerFilePath).toString()).toMatchSnapshot();
    expect(fs.readFileSync(config1.authorizerFilePath).toString()).toMatchSnapshot();
    expect(
      fs.readFileSync('testing/results/multiple/graphql/authorizers/queries/multiple/index.ts').toString()
    ).toMatchSnapshot();
    expect(
      fs.readFileSync('testing/results/multiple/graphql/authorizers/queries/index.ts').toString()
    ).toMatchSnapshot();
    expect(fs.readFileSync('testing/results/multiple/graphql/authorizers/index.ts').toString()).toMatchSnapshot();
    expect(fs.readFileSync(config0.resolverFilePath).toString()).toMatchSnapshot();
    expect(fs.readFileSync(config1.resolverFilePath).toString()).toMatchSnapshot();
    expect(fs.readFileSync(config0.resolverTestFilePath).toString()).toMatchSnapshot();
    expect(fs.readFileSync(config1.resolverTestFilePath).toString()).toMatchSnapshot();
    expect(
      fs.readFileSync('testing/results/multiple/graphql/resolvers/queries/multiple/index.ts').toString()
    ).toMatchSnapshot();
    expect(fs.readFileSync('testing/results/multiple/graphql/resolvers/queries/index.ts').toString()).toMatchSnapshot();
  });
});
