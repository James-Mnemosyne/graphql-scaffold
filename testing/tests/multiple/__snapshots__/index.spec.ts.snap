// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query with only basic object 1`] = `
"import { AuthenticationContextType } from 'src/graphql-retail/middleware/authenticate/types';

/**
 * Authorizer for thing0 query
 */
export async function thing0(context: AuthenticationContextType) {
  // DEAR CODE REVIEWER,
  // IF YOU ARE SEEING THIS, YOU SHOULD NOT ALLOW THIS THROUGH CODE REVIEW.
  // THIS IS AN AUTO-GENERATED FUNCTION
  // AND SHOULD BE REPLACED WITH VALID AUTHENTICATION LOGIC.
};
"
`;

exports[`Query with only basic object 2`] = `
"import { AuthenticationContextType } from 'src/graphql-retail/middleware/authenticate/types';

/**
 * Authorizer for thing1 query
 */
export async function thing1(context: AuthenticationContextType) {
  // DEAR CODE REVIEWER,
  // IF YOU ARE SEEING THIS, YOU SHOULD NOT ALLOW THIS THROUGH CODE REVIEW.
  // THIS IS AN AUTO-GENERATED FUNCTION
  // AND SHOULD BE REPLACED WITH VALID AUTHENTICATION LOGIC.
};
"
`;

exports[`Query with only basic object 3`] = `
"export * from './thing0';
export * from './thing1';
"
`;

exports[`Query with only basic object 4`] = `
"export * from './multiple';
"
`;

exports[`Query with only basic object 5`] = `
"export * from './queries';
"
`;

exports[`Query with only basic object 6`] = `
"import { Queries } from 'src/graphql-retail/types';

/**
 * thing0 query
 */
export const thing0: Queries['thing0'] = async function (
  _parent,
  input,
  _context
) {
  // TODO implement
  console.log('input', input);
  throw new Error('Unsupported query thing0.')
};
"
`;

exports[`Query with only basic object 7`] = `
"import { Queries } from 'src/graphql-retail/types';

/**
 * thing1 query
 */
export const thing1: Queries['thing1'] = async function (
  _parent,
  input,
  _context
) {
  // TODO implement
  console.log('input', input);
  throw new Error('Unsupported query thing1.')
};
"
`;

exports[`Query with only basic object 8`] = `
"import { mockGraphQLResolveContext, mockGraphQLResolveInfo } from 'src/graphql-retail/testing/fixtures/graphql';
import { thing0 } from '.';

/**
 * Tests for thing0 query
 */
describe('thing0 query', () => {
  beforeAll(async () => {});

  it('Runs on null input.', async () => {
    const response = await thing0(
      {},
      {}, // This is probably not valid input.
      mockGraphQLResolveContext,
      mockGraphQLResolveInfo
    );
    expect(response).toEqual(null);
  });
});
"
`;

exports[`Query with only basic object 9`] = `
"import { mockGraphQLResolveContext, mockGraphQLResolveInfo } from 'src/graphql-retail/testing/fixtures/graphql';
import { thing1 } from '.';

/**
 * Tests for thing1 query
 */
describe('thing1 query', () => {
  beforeAll(async () => {});

  it('Runs on null input.', async () => {
    const response = await thing1(
      {},
      {}, // This is probably not valid input.
      mockGraphQLResolveContext,
      mockGraphQLResolveInfo
    );
    expect(response).toEqual(null);
  });
});
"
`;

exports[`Query with only basic object 10`] = `
"export * from './thing0';
export * from './thing1';
"
`;

exports[`Query with only basic object 11`] = `
"export * from './multiple';
"
`;
